### PR TITLE
Add Involutive Ring

### PR DESCRIPTION
**Pull request by ocramz**
Introduces `adj` for denoting Complex conjugation, conjugate transpose and Hermitian adjoint (and a bunch of other things too)

### CODE REVIEW

**Comment by tonyday567** [comment]
Why is the Real constraint sneaking in there?

**Comment by ocramz** [comment]
Eh, that was introduced for excluding Complex. Or should I have used a NumHask class?

**Comment by tonyday567** [comment]
Those are some awesome laws!

It looks like Int, Integer, Float and Double will pass them.  I'd rather we just went with specific instances for these four.  One day, will have to deal with the 63 other numeric types in base but let's not do that today!

I'm a bit shy about the overlapping instance pragma as well, but that's mostly because I dont understand the implications. 

So far the library avoids Num, Fractional and Real from base.  Real in base is a hodgepodge - means trig and exponential and who knows what else.

File: numhask/src/NumHask/Algebra/Ring.hs [hunk 1]
```diff
@@ -97,3 +103,39 @@ class (Semiring a) => StarSemiring a where
 class (StarSemiring a, AdditiveIdempotent a) => KleeneAlgebra a
 
 
+
+-- | Involutive Ring
+--
+-- > adj (a + b) ==> adj a + adj b
+-- > adj (a * b) ==> adj a * adj b
+-- > adj one = one
+-- > adj (adj a) = a
```
Comment (ocramz) [inline numhask/src/NumHask/Algebra/Ring.hs]: typo: `=` should be `==>`

**Comment by tonyday567** [comment]
Just tried to checkout locally and edit this PR: result was epic fail.  Plan b: merge and work on a new PR ...

**Pull request merged**